### PR TITLE
Native invocation using `ArgumentList`

### PR DESCRIPTION
**Pull request by JamesWTruher**
This adds a new behavior for native executables which uses ArgumentList as the command arguments for the native executable.

# PR Summary

The experimental feature PSNativeCommandArgumentPassing when enabled will use the `ArgumentList` property of the `StartProcessInfo` object rather than our current mechanism of reconstructing a string when invoking a native executable.
`$PSNativeCommandArgumentPassing` is a new automatic variable which allows for runtime selection of behavior. `$PSNativeCommandArgumentPassing` may be set to either `Legacy` or `Standard`, `Legacy` is the historic behavior. The default when the experimental feature is enabled is the new `Standard` behavior. 

This new behavior *is a breaking change* from current behavior. This may break scripts and automation which work-around the various issues when invoking native applications. Historically, quotes must be escaped and it is not possible to provide empty arguments to a native application. This PR attempts to address these issues.

new behaviors made available by this change

- literal or expandable strings with embedded quotes the quotes are now preserved
```powershell
PS > $a = 'a" "b'
PS > $PSNativeCommandArgumentPassing = "Legacy"
PS > testexe -echoargs $a 'a" "b' a" "b    
Arg 0 is <a b>
Arg 1 is <a b>
Arg 2 is <a b>
PS > $PSNativeCommandArgumentPassing = "Standard"
PS > testexe -echoargs $a 'a" "b' a" "b    
Arg 0 is <a" "b>
Arg 1 is <a" "b>
Arg 2 is <a b>
```

- empty strings as arguments are now preserved:
```powershell
PS>  $PSNativeCommandArgumentPassing = "Legacy"
PS> testexe -echoargs '' a b ''           
Arg 0 is <a>
Arg 1 is <b>
PS> $PSNativeCommandArgumentPassing = "Standard"
PS> testexe -echoargs '' a b ''           
Arg 0 is <>
Arg 1 is <a>
Arg 2 is <b>
Arg 3 is <>
```

the new behavior does not change invocations which look like this:

```powershell
PS> $PSNativeCommandArgumentPassing = "Legacy"                                            
PS> testexe -echoargs -k com:port=\\devbox\pipe\debug,pipe,resets=0,reconnect          
Arg 0 is <-k>
Arg 1 is <com:port=\\devbox\pipe\debug,pipe,resets=0,reconnect>
PS> $PSNativeCommandArgumentPassing = "Standard"                                  
PS> testexe -echoargs -k com:port=\\devbox\pipe\debug,pipe,resets=0,reconnect
Arg 0 is <-k>
Arg 1 is <com:port=\\devbox\pipe\debug,pipe,resets=0,reconnect>
```

Additionally, parameter tracing is now provided so `Trace-Command` will provide useful information for debugging

```powershell
PS> $PSNativeCommandArgumentPassing = "Legacy"                                           
PS> trace-command -PSHOST -Name ParameterBinding { testexe -echoargs $a 'a" "b' a" "b }
DEBUG: 2021-02-01 17:19:53.6438 ParameterBinding Information: 0 : BIND NAMED native application line args [/Users/james/src/github/forks/jameswtruher/PowerShell-1/test/tools/TestExe/bin/testexe]
DEBUG: 2021-02-01 17:19:53.6440 ParameterBinding Information: 0 :     BIND argument [-echoargs a" "b a" "b "a b"]
DEBUG: 2021-02-01 17:19:53.6522 ParameterBinding Information: 0 : CALLING BeginProcessing
Arg 0 is <a b>
Arg 1 is <a b>
Arg 2 is <a b>
PS> $PSNativeCommandArgumentPassing = "Standard"                                            
PS> trace-command -PSHOST -Name ParameterBinding { testexe -echoargs $a 'a" "b' a" "b }
DEBUG: 2021-02-01 17:20:01.9829 ParameterBinding Information: 0 : BIND NAMED native application line args [/Users/james/src/github/forks/jameswtruher/PowerShell-1/test/tools/TestExe/bin/testexe]
DEBUG: 2021-02-01 17:20:01.9829 ParameterBinding Information: 0 :     BIND cmd line arg [-echoargs] to position [0]
DEBUG: 2021-02-01 17:20:01.9830 ParameterBinding Information: 0 :     BIND cmd line arg [a" "b] to position [1]
DEBUG: 2021-02-01 17:20:01.9830 ParameterBinding Information: 0 :     BIND cmd line arg [a" "b] to position [2]
DEBUG: 2021-02-01 17:20:01.9831 ParameterBinding Information: 0 :     BIND cmd line arg [a b] to position [3]
DEBUG: 2021-02-01 17:20:01.9908 ParameterBinding Information: 0 : CALLING BeginProcessing
Arg 0 is <a" "b>
Arg 1 is <a" "b>
Arg 2 is <a b>
```

## PR Context

Invoking native applications is sometimes very problematic because of how the application is invoked. A new property is available on the `ProcessStartInfo` object which hopes to improve the experience. This PR takes advantage of this new property.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [x] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): PSNativeCommandArgumentPassing
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/PowerShell/PowerShell/issues/13068
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [x] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

The impact on completions may be affected if those completers are calling native applications.

### CODE REVIEW

**Comment by mklement0** [comment]
* While switching to `.ArgumentList` is definitely called for and fully solves the problem _on Unix-like platforms_, _on Windows_ there are _also_ vital accommodations that we need to make for common use cases, notably calling _batch files_ and _`misexec`-style CLIs_, as detailed in [this comment](https://github.com/PowerShell/PowerShell/issues/1995#issuecomment-738073284).

  * If there's disagreement about these accommodations, let's have a discussion, but none has happened yet (except with other community members), so, yes, perhaps an RFC is necessary.

* Both improvements should be part of the _same_ PR, to once and for all solve 100% of all argument-passing problems on Unix, and 99% of all problems on Windows (the remaining problems cannot be fixed generically, and must be addressed by users via `--%`).

* That is, only _one_ (opt-in) breaking change should be made, and it should be controlled by a _single_ preference variable (which is the next best solution if the breaking change is deemed unacceptable).

* As for the preference variable:

  * First, the name `$PSNativeApplicationUsesArgumentList` is too technical to begin with (end users who expect argument-passing in their shell to "just work" cannot be expected not know about plumbing details such as `.Arguments` vs. `.ArgumentList`, which would make discovery difficult); also, the variable should be `[bool]`-typed (while assigning `1` works through implicit conversion, we should show `$true` / `$false` in the docs).

  * Second, with the PR encompassing the Windows-only accommodations too, a broader name is called for. I'm struggling to come up with a good suggestion. `$PSFixNativeArgumentPassing`?

**Comment by SteveL-MSFT** [comment]
For the variable, I prefer `$PSNativeCommandArgumentParsing` as an enum with values `ArgumentList`, and `String` (can't think of something better...).

**Comment by mklement0** [comment]
@SteveL-MSFT:

>  I prefer `$PSNativeCommandArgumentParsing`

Not sure if it's just a typo, but to be clear: this isn't about argument _parsing_, it's about argument _passing_ (which, on Windows only, also involves _encoding_, as a single string - which `.ArgumentList` will do for us, but the suggested accommodations will have to be coded manually).

> an enum with values `ArgumentList`, and `String` (can't think of something better...).

I can see why you would to avoid the word "fix" in the variable name - even though that's what it comes down to.

Again, what matters to the user is only the following - not _technical_ characterizations based on plumbing users may not even know of:

* Is the *old* behavior in effect, where I need to apply *workarounds*?
* Or is the *new* behavior in effect, where things work as expected?

Thus, if you do want an `enum`, I suggest a ``$PSNativeCommandArgumentPassing` variable with the following enumeration values, corresponding to the above:

* `Legacy`
* `Standard` (or similar)

**Review by Steg17 — APPROVED** [review]

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this, we agreed on automatic variable `$PSNativeCommandArgumentPassing` as an enum with values `Legacy` and `Standard` (default).  This feature needs to be an ExperimentalFeature.

File: src/System.Management.Automation/engine/NativeCommandParameterBinder.cs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 
 using System.Collections;
 using System.Collections.ObjectModel;
+using System.Collections.Generic;
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/NativeCommandParameterBinder.cs:6]: namespaces are alphabetical, so this should be above ObjectModel

File: src/System.Management.Automation/engine/NativeCommandParameterBinder.cs [hunk 2]
```diff
@@ -151,6 +152,30 @@ internal string Arguments
 
         private readonly StringBuilder _arguments = new StringBuilder();
 
+        internal string[] ArgumentList
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/NativeCommandParameterBinder.cs:155]: It doesn't appear that `ArgumentList` couldn't stay a List instead of being an array?

File: test/powershell/Language/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1 [hunk 3]
```diff
@@ -1,70 +1,84 @@
...
+    $PSNativeApplicationUsesArgumentList = $argumentListValue
+    Describe "Native Command Arguments" -tags "CI" {
+        # When passing arguments to native commands, quoted segments that contain
+        # spaces need to be quoted with '"' characters when they are passed to the
+        # native command (or to bash or sh on Linux).
+        #
+        # This test checks that the proper quoting is occuring by passing arguments
+        # to the testexe native command and looking at how it got the arguments.
+        It "Should handle quoted spaces correctly (ArgumentList=${PSNativeApplicationUsesArgumentList})" {
+            $a = 'a"b c"d'
```
Comment (SteveL-MSFT) [inline test/powershell/Language/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1:13]: which test case represents the MSI scenario on Windows and `git commit -m "This 'has' quotes"`

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/NativeCommandParameterBinder.cs (hunk 1 above)
Comment (JamesWTruher) [reply src/System.Management.Automation/engine/NativeCommandParameterBinder.cs:5]: fixed

File: src/System.Management.Automation/engine/InitialSessionState.cs [hunk 4]
```diff
@@ -4551,6 +4577,12 @@ .ForwardHelpCategory Cmdlet
                 ScopedItemOptions.None),
             // End: Variables which control remoting behavior
 
+            new SessionStateVariableEntry(
+                SpecialVariables.PSSessionApplicationName,
+                "wsman",
+                RemotingErrorIdStrings.PSSessionAppName,
+                ScopedItemOptions.None),
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/InitialSessionState.cs:4584]: This is a duplicate and should be removed.

Comment (JamesWTruher) [reply src/System.Management.Automation/engine/InitialSessionState.cs:4584]: fixed

File: test/powershell/Language/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1 (hunk 3 above)
Comment (JamesWTruher) [reply test/powershell/Language/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1:26]: i don't think that's currently broken in our current environment, but it's easy enough to add to this.

File: src/System.Management.Automation/engine/NativeCommandParameterBinder.cs (hunk 2 above)
Comment (JamesWTruher) [reply src/System.Management.Automation/engine/NativeCommandParameterBinder.cs:155]: the API takes a string array, so I thought it would be better for me to do it this rather than calling the `ToArray` inline.

File: src/System.Management.Automation/engine/ExperimentalFeature/ExperimentalFeature.cs [hunk 5]
```diff
@@ -136,6 +136,9 @@ static ExperimentalFeature()
                 new ExperimentalFeature(
                     name: PSAnsiProgressFeatureName,
                     description: "Enable lightweight progress bar that leverages ANSI codes for rendering"),
+                new ExperimentalFeature(
+                    name: "PSNativeCommandArgumentPassing",
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/ExperimentalFeature/ExperimentalFeature.cs:140]: Should define the experimental feature name as a const string, see line 24 in this file

File: src/System.Management.Automation/engine/InitialSessionState.cs [hunk 6]
```diff
@@ -4314,8 +4340,8 @@ .FORWARDHELPCATEGORY Cmdlet
             if ($pagerArgs) {
                 # Supply pager arguments to an application without any PowerShell parsing of the arguments.
                 # Leave environment variable to help user debug arguments supplied in $env:PAGER.
-                $env:__PSPAGER_ARGS = $pagerArgs
-                $help | Out-String -Stream -Width ($consoleWidth - 1) | & $pagerCommand --% %__PSPAGER_ARGS%
+                # $env:__PSPAGER_ARGS = $pagerArgs
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/InitialSessionState.cs:4343]: should probably wrap this with check for experimental feature so we can delete the env var when this feature is stable

File: test/powershell/Host/ConsoleHost.Tests.ps1 [hunk 7]
```diff
@@ -235,8 +235,8 @@ Describe "ConsoleHost unit tests" -tags "Feature" {
             $observed | Should -BeExactly "h-llo"
         }
 
-        It "Empty command should fail" {
-            & $powershell -noprofile -c ''
+        It "Missing command should fail" {
```
Comment (SteveL-MSFT) [inline test/powershell/Host/ConsoleHost.Tests.ps1:238]: Should there be a new test for empty command?

File: src/System.Management.Automation/engine/ExperimentalFeature/ExperimentalFeature.cs (hunk 5 above)
Comment (JamesWTruher) [reply src/System.Management.Automation/engine/ExperimentalFeature/ExperimentalFeature.cs:140]: fixed

File: test/powershell/Host/ConsoleHost.Tests.ps1 (hunk 7 above)
Comment (JamesWTruher) [reply test/powershell/Host/ConsoleHost.Tests.ps1:238]: `& $powershell -noprofile -c ''` is now a valid command (it returns an empty string) which is similar to the the next test, but I can add a test case specifically for an empty string.

File: src/System.Management.Automation/engine/InitialSessionState.cs (hunk 6 above)
Comment (JamesWTruher) [reply src/System.Management.Automation/engine/InitialSessionState.cs:4343]: fixed

File: src/System.Management.Automation/engine/InitialSessionState.cs [hunk 8]
```diff
@@ -4298,7 +4298,13 @@ .FORWARDHELPCATEGORY Cmdlet
         }
         else {
             $pagerCommand = 'less'
-            $pagerArgs = '-s','-P','Page %db?B of %D:.\. Press h for help or q to quit\.'
+            # PSNativeCommandArgumentPassing arguments should be constructed differently.
+            if ((Get-ExperimentalFeature PSNativeCommandArgumentPassing).Enabled) {
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/InitialSessionState.cs:4302]: You can use `$EnabledExperimentalFeatures` and avoid the cmdlet call

File: test/powershell/Language/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1 [hunk 9]
```diff
@@ -1,70 +1,98 @@
...
+            @{arguments = "1 ,2,3"; expected = "1", ",2,3"}
+            @{arguments = "1 , 2,3"; expected = "1", ",", "2,3"}
+        ) {
+            param($arguments, $expected)
+            $lines = @(Invoke-Expression "testexe -echoargs $arguments")
+            $lines.Count | Should -Be $expected.Count
+            for ($i = 0; $i -lt $expected.Count; $i++) {
+                $lines[$i] | Should -BeExactly "Arg $i is <$($expected[$i])>"
+            }
         }
```
Comment (SteveL-MSFT) [inline test/powershell/Language/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1:93]: Should have explicit test for equivalent to: `exe /c:"some value"`

File: src/System.Management.Automation/engine/InitialSessionState.cs (hunk 8 above)
Comment (JamesWTruher) [reply src/System.Management.Automation/engine/InitialSessionState.cs:4302]: fixed

File: test/powershell/Language/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1 (hunk 9 above)
Comment (JamesWTruher) [reply test/powershell/Language/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1:100]: fixed

File: test/powershell/Language/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1 [hunk 10]
```diff
@@ -1,70 +1,112 @@
...
+        # spaces need to be quoted with '"' characters when they are passed to the
+        # native command (or to bash or sh on Linux).
+        #
+        # This test checks that the proper quoting is occuring by passing arguments
+        # to the testexe native command and looking at how it got the arguments.
+        It "Should handle quoted spaces correctly (ArgumentList=${PSNativeCommandArgumentPassing})" {
+            $a = 'a"b c"d'
+            $lines = testexe -echoargs $a 'a"b c"d' a"b c"d "a'b c'd"
+            $lines.Count | Should -Be 4
+            if ( (Get-ExperimentalFeature PSNativeCommandArgumentPassing).Enabled -and $PSNativeCommandArgumentPassing -ne "Legacy" ) {
```
Comment (SteveL-MSFT) [inline test/powershell/Language/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1:29]: There's still a bunch of calling `Get-ExperimentalFeature` that can be replaced with `$EnabledExperimentalFeatures`

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

**Review by SteveL-MSFT — APPROVED** [review]
LGTM

**Comment by TravisEz13** [comment]
@daxian-dbw  Can you update your review?

File: src/System.Management.Automation/engine/NativeCommandParameterBinder.cs [hunk 11]
```diff
@@ -187,7 +250,8 @@ private void AppendOneNativeArgument(ExecutionContext context, object obj, Array
                         break;
                     }
 
-                    arg = PSObject.ToStringParser(context, ParserOps.Current(null, list));
+                    currentObj = ParserOps.Current(null, list);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/NativeCommandParameterBinder.cs:253]:
```suggestion
                    currentObj = ParserOps.Current(IScriptExtent: null, list);
```

File: src/System.Management.Automation/engine/NativeCommandParameterBinder.cs [hunk 12]
```diff
@@ -198,12 +262,16 @@ private void AppendOneNativeArgument(ExecutionContext context, object obj, Array
...
+                    // Only add the separator to the argument string rather than adding a separator to the ArgumentList.
                     _arguments.Append(separator);
 
                     if (sawVerbatimArgumentMarker)
                     {
                         arg = Environment.ExpandEnvironmentVariables(arg);
                         _arguments.Append(arg);
+
+                        // we need to split the argument on spaces
+                        _argumentList.AddRange(arg.Split(" ", StringSplitOptions.RemoveEmptyEntries));
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/NativeCommandParameterBinder.cs:274]:
```suggestion
                        _argumentList.AddRange(arg.Split(' ', StringSplitOptions.RemoveEmptyEntries));
```

File: src/System.Management.Automation/engine/NativeCommandParameterBinder.cs [hunk 13]
```diff
@@ -244,10 +314,28 @@ private void AppendOneNativeArgument(ExecutionContext context, object obj, Array
                         }
                         else
                         {
-                            PossiblyGlobArg(arg, stringConstantType);
+                            if (argArrayAst != null && UseArgumentList)
+                            {
+                                // We have a literal array, so take the extent, break it on spaces and add them to the argument list.
+                                foreach (string element in argArrayAst.Extent.Text.Split(" ", StringSplitOptions.RemoveEmptyEntries))
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/NativeCommandParameterBinder.cs:320]:
```suggestion
                                foreach (string element in argArrayAst.Extent.Text.Split(' ', StringSplitOptions.RemoveEmptyEntries))
```

File: src/System.Management.Automation/engine/NativeCommandParameterBinder.cs (hunk 11 above)
Comment (JamesWTruher) [reply src/System.Management.Automation/engine/NativeCommandParameterBinder.cs:253]: uh, this doesn't actually compile.

File: src/System.Management.Automation/engine/InitialSessionState.cs [hunk 14]
```diff
@@ -1413,7 +1439,7 @@ private static InitialSessionState CreateRestrictedForRemoteServer()
             }
 
             // Add built-in variables.
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/InitialSessionState.cs:1441]: `LocalRunspace.ResetRunspaceState()` adds all variables defined in `BuiltInVariables` when resetting the Runspace, which bypass the `AddVariables` method. However, we can probably ignore this for this PR, because I think that behavior of `ResetRunspaceState` is not correct. I opened #14987 for it.

File: src/System.Management.Automation/engine/InitialSessionState.cs [hunk 15]
```diff
@@ -4312,10 +4344,16 @@ .FORWARDHELPCATEGORY Cmdlet
...
+                # Otherwise, supply pager arguments to an application without any PowerShell parsing of the arguments.
                 # Leave environment variable to help user debug arguments supplied in $env:PAGER.
-                $env:__PSPAGER_ARGS = $pagerArgs
-                $help | Out-String -Stream -Width ($consoleWidth - 1) | & $pagerCommand --% %__PSPAGER_ARGS%
+                if ($EnabledExperimentalFeatures -contains 'PSNativeCommandArgumentPassing') {
+                    $help | Out-String -Stream -Width ($consoleWidth - 1) | & $pagerCommand $pagerArgs
+                }
+                else {
+                    $env:__PSPAGER_ARGS = $pagerArgs
+                    $help | Out-String -Stream -Width ($consoleWidth - 1) | & $pagerCommand --% %__PSPAGER_ARGS%
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/InitialSessionState.cs:4355]: > & $pagerCommand --% %__PSPAGER_ARGS%

It looks we will get here only if on Unix. `%__PSPAGER_ARGS%` is expanded by PowerShell, right?

**Review by daxian-dbw — APPROVED** [review]

File: src/System.Management.Automation/engine/InitialSessionState.cs (hunk 15 above)
Comment (JamesWTruher) [reply src/System.Management.Automation/engine/InitialSessionState.cs:4355]: it should reach this code when the pager is a standalone executable (regardless of platform) PowerShell expands this (with `Environnment.ExpandEnvironmentVariable` before passing it to the native app.

**Comment by PoshChan** [comment]
@TravisEz13, this is the reminder you requested 1 hour ago

**Comment by iSazonov** [comment]
Since this PR introduces a _breaking_ change anyway and is being implemented as an _experimental_ feature, I would like to understand why don't we implement mklement0's proposal in full as an experimental feature and _close the long history_?
See:
- https://github.com/PowerShell/PowerShell/pull/14692#issuecomment-772816850
- https://github.com/PowerShell/PowerShell/issues/1995#issuecomment-738073284
- https://github.com/PowerShell/PowerShell/issues/1995#issuecomment-738073284

Also I asked previously and still wonder why @mklement0 is not in a PowerShell Committee.  I believe MSFT team would have saved a lot of time and effort in this case.

**Comment by TravisEz13** [comment]
@iSazonov  can you open a new issue w.r.t. https://github.com/PowerShell/PowerShell/pull/14692#issuecomment-811911407 and mark it for committee review?

**Comment by TravisEz13** [comment]
@PoshChan Please remind me in 1 Hour

**Comment by mklement0** [comment]
The (hopefully) final proposal is now in https://github.com/PowerShell/PowerShell/issues/14747#issuecomment-782892238

**Comment by PoshChan** [comment]
@TravisEz13, this is the reminder you requested 1 Hour ago
